docs: Improve ORAS diagnose experience (#1483)
oras-project/oras: docs/proposals/diagnose-experience.md

## Concepts

At first, the output of ORAS flag `--verbose` and `--debug` should be clarified before restructuring them. 

### Output

There are four types of output in ORAS CLI:

- **Status output**: such as progress information, progress bar in pulling or pushing files.
- **Metadata output**: showing what has been pulled (e.g. filename, digest, etc.) in specified format, such as JSON, text.
- **Content output**: it is to output the raw data obtained from the remote registry server or file system, such as the pulled artifact content save as a file.
- **Error output**: error message are expected to be helpful to troubleshoot where the user has done something wrong and the program is guiding them in the right direction.

The target users of these types of output are general users. 

Currently, the output of ORAS `--verbose` flag only exists in oras `pull/push/attach/discover` commands, which prints out detailed status output and metadata output. Since ORAS v1.2.0, progress bar is enabled in `pull/push/attach` by default, thus the ORAS output is already verbose on a terminal.

> It is intended for end-users who want to observe the detailed file operation when using ORAS. It gives users a comprehensive view of what the tool is doing at every step and how long does it take when push or pull a file.

Author: It refers to the original output of ORAS `--verbose`

  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`

### **Avoid Logging Sensitive Information**
- **Privacy and Security:** Abstain from logging sensitive information such as passwords, personal data, or authentication tokens.
  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]`  (exclude authentication token and password information).

## Proposals for ORAS CLI

Based on the concepts above, 

- Deprecate the global flag `--verbose` and only remain `--debug` to avoid ambiguity. Based on the concept above, it is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs in ORAS as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.
- Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 

> - Add an empty line as the separator between each request and response for readability.

Reviewer: **An** empty line or **some** empty lines?

Author: I changed to 2 lines

- Add timestamp of each request and response to the beginning of each request and response.

> - Add the response body including [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and the metadata of processed OCI object (e.g. image manifest) to the debug logs.

Reviewer 2: We should never print out binary response body. We need a list of printable content types such as `application/json`.

Author: I revised this convention to:

> Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.

> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew"
   "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0001] Response #0
< Response Status: "401 Unauthorized"
< Response headers:
   "Content-Length": "73"
   "X-Github-Request-Id": "9FC6:30019C:17C06:1C462:66AD0463"
   "Content-Type": "application/json"
   "Www-Authenticate": "Bearer realm=\"https://ghcr.io/token\",service=\"ghcr.io\",scope=\"repository:oras-project/oras:pull\""
   "Date": "Fri, 02 Aug 2024 16:08:04 GMT" 
DEBU[0001] Request #1
> Request URL: "https://ghcr.io/token?scope=repository%3Aoras-project%2Foras%3Apull&service=ghcr.io"
> Request method: "GET"
> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew" 
DEBU[0002] Response #1
< Response Status: "200 OK"
< Response headers:
   "Content-Type": "application/json"
   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 

> ```

Author: Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, expect for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on Terminal but keeping plain text in a non-TTY mode respectively.

**Suggested debug logs:**

```

> 2024-08-02 23:56:02 > msg=Request #0

Reviewer: The example is not consistent with the one proposed in the **Logs** section. How about something like

```
[2024-10-10T19:26:33+08:00] --> Request #0
> Request URL: "http://localhost:5000/v2/240923test1/manifests/v1"
> Request method: "HEAD"

[2024-10-10T19:26:33+08:00] <-- Response #0
> Request URL: "http://localhost:5000/v2/240923test1/manifests/v1"
> Request method: "HEAD"
```

for both TTY and non-TTY?

Author: Looks good. Making both TTY and non-TTY consistent would be the right direction. I changed the example accordingly

> Request URL: "https://ghcr.io/v2/oras-project/oras/manifests/v1.2.0"
> Request method: "GET"
> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew"
   "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 

2024-08-02 23:56:03 <  msg=Response #0
< Response Status: "401 Unauthorized"
< Response headers:
   "Content-Length": "73"
   "X-Github-Request-Id": "9FC6:30019C:17C06:1C462:66AD0463"
   "Content-Type": "application/json"
   "Www-Authenticate": "Bearer realm=\"https://ghcr.io/token\",service=\"ghcr.io\",scope=\"repository:oras-project/oras:pull\""
   "Date": "Fri, 02 Aug 2024 23:56:03 GMT"
< Response body:

> {

Reviewer 2: The format seems wrong

Uploading 655f5cc5d5d2 application/vnd.oci.image.manifest.v1+json
Uploaded  655f5cc5d5d2 application/vnd.oci.image.manifest.v1+json
Pushed [oci-layout] layout-test:sample
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:655f5cc5d5d2e7ff2ab90378514103996523b065ce5ef43cd6e6a4de7d80a535
```

### Show user's environment details

Output the user's running environment details of ORAS such as operating system and architecture information would be helpful to help the ORAS developers locate the issue and reproduce easier. 

For example, the operating system and architecture are supposed to be outputted in `oras version`: 

```bash
$ oras version

ORAS Version:    1.2.0+Homebrew
Go version:      go1.22.3
OS/Arch:         linux/amd64
Git commit:      xxxxxxxxxxxx
Git tree state:  clean
```

## Q & A

**Q1:** Is it a common practice to use an environment variable like export ORAS_DEBUG=1 as a global switch for debug logs? What are the Pros and Cons of using this design?

**A:** Per our discussion in the ORAS community meeting, ORAS maintainers agreed to not introduce an additional environment variable as a global switch to enable debug logs since --debug is intuitive enough.

> **Q2:**: For the diagnose flag options, why deprecate `--verbose` and remain `--debug` as it is?

Reviewer: nit: duplicated `:`

# Improve ORAS diagnose experience

> ORAS currently offers two global options, `--verbose` and `--debug`, which enable users to generate verbose output and debug logs respectively. These features facilitate both users and developers in inspecting ORAS's performance, interactions with external services and internal systems, and in diagnosing issues by providing a clear picture of the tool’s operations.

Reviewer 3: Why not call verbose log directly, is there any difference between verbose log and verbose ouput?

> ORAS currently offers two global options, `--verbose` and `--debug`, which enable users to generate verbose output and debug logs respectively. These features facilitate both users and developers in inspecting ORAS's performance, interactions with external services and internal systems, and in diagnosing issues by providing a clear picture of the tool’s operations.

Author: I was thinking to reference `--verbose` output and the target user is for normal users.

Given the diverse roles and scenarios in which ORAS CLI is utilized, we have received feedback from users and developers to improve the diagnostic experience. Enhancing the verbose output and debug logs can significantly benefit ORAS users and developers by making diagnostics clearer and more unambiguous.

This proposal document aims to:

1. Identify the issues associated with the current implementation of the `--verbose` and `--debug` options.
2. Clarify the concepts of verbose output and debug logs.
3. List the guiding principles to write comprehensive, clear, and conducive verbose output and debug logs for effective diagnosis.
4. Propose solutions to improve the diagnostic experience for ORAS CLI users and developers.

## Problem Statement

> Specifically, there are exiting GitHub issues raised in the ORAS community.

Reviewer 2: ```suggestion
Specifically, there are existing GitHub issues raised in the ORAS community.
```

- The user is confused about when to use `--verbose` and `--debug`. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).
- Poor readability of debug logs. No separator lines between request and response information. Users need to add separator lines manually for readability. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).

> - Critical information is missing in debug logs. For example, the [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and metadata of the processed resource object (e.g. image manifest) are not displayed.

Reviewer 4: The term "resource object" is unfamiliar to me. Is this defined somewhere? The clarification in parentheses mentions "image manifest" but are "resource objects" "image manifests" only or also other "resources" or "objects"? IMHO, the term is confusing.

Author: I changed to `objects` since Kubernetes has a similar term https://kubernetes.io/docs/concepts/overview/working-with-objects/

- The detailed operation information is missing in verbose output. For example, how many resource objects are processed where. Less or none verbose output of ORAS commands in some use cases.
- Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

> ## Concepts

Reviewer 2: In terms of concepts, a more scientific classification would be

- Output
  - Status output (e.g. progress bar)
  - Metadata output (e.g. json output)
  - Content output (e.g. raw manifest fetched from remote)
- Log
  - Debug log (for diagnostics)
  - Information log (not used by `oras`)
  - Warning log (not used by `oras`)
  - Error log (not used by `oras`)

Note:
- Information log should not be confused with status output or metadata output.
- There is no concept of "verbose log/output". The term "verbose log/output" can be interpreted as "not simplified output". If the output cannot be simplified, then there is no "verbose output" at all.

Reviewer: What does simplified output include?

Author: This is an valid classification if we split this concept into output and log. We had a related discussion when designing the formatted output https://hackmd.io/WIpCbRcUR36v_Wfqb0NicQ

Author: For the diagnose flag options, if only DEBUG level log is captured and generated by ORAS, is it much more straightforward and less breaking if we only deprecate `--verbose` and remain `--debug` as it is?

Author: Updated the concepts with @shizhMSFT 's suggestions.

Before restructuring the log information of ORAS debug log and [verbose output](https://en.wikipedia.org/wiki/Verbose_mode), it worth clarifying the concepts and differences between verbose output and debug logs.

> ### Verbose Output 

Reviewer 2: I would like to discuss the reason why we have "verbose output" in the first place.

In my understanding, there is no concept of "verbose output" in the first place. The CLI prints as much stuffs as possible for users to understand. However, it prints too much and distracts the user. Therefore, we reduce the output a bit. The reduced output is now considered as "standard output", and the original output is considered as "verbose output".

Author: Refactor the concept section and removed the concept of "verbose output"

Verbose output focuses on providing a comprehensive, high-level view of the application's operations. It is intended for end-users who want to observe the detailed normal operation of the tool. It should be human readable and descriptive.

- **Purpose**: Verbose output are typically used to provide highly detailed information about the flow and operations of the application. They give users a comprehensive view of what the tool is doing at every step.

> - **Target users**: These are generally intended for users who want to understand the detailed workings of the tool, not necessarily for debugging specific issues.

Reviewer 5: +1. A user trying to copy an artifact might experience failure and is trying to debug their failure. Hence the tool should provide verbose logs to help debug scenarios and gain insight into the overall flow of operations.

> - **Target users**: These are generally intended for users who want to understand the detailed workings of the tool, not necessarily for debugging specific issues.

Author: Per our community meeting discussion, ORAS maintainers agreed to deprecate `--verbose` and remain `--debug` as a unified control of debug logs. I updated the spec to avoid ambiguity.

- **Content**: Verbose output includes lots of informational messages about the application's state, operations performed, configuration details, and more. They provide a broader view, which is helpful for tracing the overall execution of the tool.

> - **Level of Detail**: Very detailed, but usually focused on normal operations rather than errors or issues.

Reviewer: >  If things are working well, people will not use verbose output.

Not really, sometimes people just want to understand what's happening behind the scenes, or just to play around.

> - **Level of Detail**: Very detailed, but usually focused on normal operations rather than errors or issues.

Author: I removed the "verbose output" concept and clarified different types of output in ORAS. PTAL https://github.com/oras-project/oras/pull/1483/files#diff-aa5e00cae2789dcb86374bcad7c8469a3cd6f7bcff09e4e0a69091aa4c692048R28

### Debug Logs

> Debug logs focus on providing technical details for in-depth diagnosing and troubleshooting issues. It is intended for developers or technical users who need to understand the inner workings of the tool. Debug logs are detailed and technical, often including HTTP request and response from interactions between client and server, as well as code-specific information.

Author: HTTP trace information should be included in debug logs

> - **Purpose**: Debug logs are specifically aimed at helping developers diagnose and fix issues within the application. They contain detailed technical information that is useful for troubleshooting problems.

Reviewer 5: The term application needs context. If its a user, their app might be a script or a GH action making a call to the tool. 
Debug logs in ORAS's case in my opinion should target submitting bug reports to ORAS incase of failed behavior. 
So I would say the goal is to facilitate ORAS developers to fix or diagnose ORAS. Not for a tool user to fix their application.

Author: Agree. Updated

> - **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of issues.

Reviewer 5: ```suggestion
- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of a possible issue with the tool itself.
```

- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.

## Guiding Principles

> Here are the guiding principles to print out debug logs.

Reviewer 3: Guiding principles for debug logs not for verbose logs/output?

### 1. **Timestamp Each Log Entry**
- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.
  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`

### 2. **Capture API-Specific Details**

> - **API Requests:** Log detailed information about API requests made to the registry server, including the HTTP method, endpoint, headers, and body (excluding sensitive information).

Author: Will it be insecure and non-compliant to log sensitive information in ORAS debug logs?

  - Example: `DEBUG: [HTTPRequest] POST /v2/oras-demo/blobs/uploads/ Headers: {Content-Length: 524288}`
  
- **API Responses:** Log details about the API responses received, including status codes, headers, and response body (excluding sensitive information).
  - Example: `DEBUG: [HTTPResponse] Status: 201 Created, Headers: {Location: /v2/oras-demo/blobs/uploads/uuid}, Body: {}`

### 3. **Log Before and After Critical Operations**
- **Operation Logs:** Log before performing critical operations and after completing them, including success or failure status.
  - Example: `DEBUG: Starting upload of layer 2 of 3 for repository oras-demo`

>   - Example: `DEBUG: Successfully uploaded layer 2 of 3 for repository oras-demo`

Reviewer: I thought this is already covered by the progress bar

Author: Yes, indeed.

- **State Logs:** Log important state information before and after key operations or decisions.

>   - Example: `DEBUG: Current retry attempt: 1, Max retries: 3`

Reviewer: Do we also want to mention function names, important variable state, etc?

Author: ACK. Updated

### 4. **Error and Exception Handling**
- **Catch and Log Exceptions:** Always catch exceptions and log them with relevant context and stack traces.

>   - Example: `ERROR: Exception occurred in fetchManifest: Network timeout while accessing /v2/oras-demo/manifests/latest`

Reviewer 5: Why is this not in the verbose logs, if an upstream is experience connection failures users should know about this in verbose logs.

Author: This should be included in the debug logs enabled by `--debug` per ORAS meeting discussion https://hackmd.io/P-O6n222TcSMoJgHmTTduw?view#Notes

  
- **Error Codes:** Include specific error codes to facilitate quick identification and resolution.
  - Example: `ERROR: [ErrorCode: 504] Network timeout while accessing /v2/oras-demo/manifests/latest`

### 5. **Environment and Configuration Details**
- **Environment Information:** Log details about the environment where the tool is running, such as OS version and architecture, tool version, and environment variables (excluding sensitive data).

>   - Example: `DEBUG: Running on OS: Ubuntu 20.04, Arch: Linux AMD64 ORAS version: v1.2.0`

Reviewer: Where to output this? Should we just put this information in `oras version` like docker does?

Reviewer 5: +1 could be an enhancement to version like 
`oras version --long` or another similar flag.

Author: Per our [community meeting discussion](https://hackmd.io/P-O6n222TcSMoJgHmTTduw?view#Notes), only the operating system and architecture are supposed to be outputted in `oras version`:

```bash
$ oras version

ORAS Version:    1.2.0+Homebrew
Go version: go1.22.3
OS/Arch: Linux AMD64
```

> - Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, expect for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on Terminal but keeping plain text in a non-TTY mode respectively.

Reviewer 6: ```suggestion
- Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on Terminal but keeping plain text in a non-TTY mode respectively.
```

- **Configuration Settings:** Include current configuration settings that might affect the execution or behavior of the tool.
  - Example: `DEBUG: Configuration - Registry URL: https://myregistry.io, Timeout: 30s, RetryCount: 3`

### 5. **Include Actionable Information**
- **Guidance:** Where possible, provide suggestions within logs about potential fixes or next steps for resolving issues.
  - Example: `DEBUG: [Action] Check network connectivity and retry the operation. Error occurred while accessing /v2/oras-demo/manifests/latest`

> - **Diagnostic Tips:** Include information that can assist in diagnosing issues like configuration settings, environment variables, or system states.

Reviewer: What is the difference between this and **5. Environment and Configuration Details**?

Author: I removed this point as it is duplicated.

>   - Example: `DEBUG: Current registry URL: https://myregistry.io, Timeout setting:

Reviewer: nit: broken format

### 6. **Avoid Logging Sensitive Information**
- **Privacy:** Abstain from logging sensitive information such as passwords, personal data, or security tokens.
  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]` (without password details)

- **Compliance:** Ensure that logs adhere to relevant data protection and privacy regulations.
  - Example: Anonymize or obfuscate sensitive customer data in logs.

> ## Proposals for ORAS CLI

Reviewer 5: +1 and recommend upfront a summary and then dive into the changes.

> ## Proposals for ORAS CLI

Author: Per our community meeting discussion, ORAS maintainers agreed to deprecate `--verbose` and remain `--debug` as a unified control of debug logs. 

The major reason is that this change avoids overloading the flag `--verbose` and reduce ambiguity in ORAS diagnose experience. Moreover, the `--debug` is consistent with other popular container client tools, such as Helm and Docker. Deprecation of `--verbose` is less breaking than changing behaviors of `--debug`.

- Deprecate the global flag `--debug` and only remain `--verbose` to avoid ambiguous usage. 

> - Make the debug logs as an optional output controlled by `--verbose` via a parameter. Debug logs should be sent to stderr. 

Reviewer 4: So, we are saying that we will still have two parameters? Is this correct? Or will the command parameter be `--verbose more-verbose`? This is a bit confusing how will it work.

Reviewer 6: Are there going to be levels of debug/verbosity? It seems to me we are talking more about debug output rather than verbosity.

Reviewer 5: Recommend introducing instead of any flag and the environment variable is totally not supported. 
```
export ORAS_DEBUG=1
oras pull ...
```

Author: Per [community meeting discussion on Sep 3](https://hackmd.io/P-O6n222TcSMoJgHmTTduw?view#Notes), ORAS maintainers agreed to not introduce an additional environment variable as a global switch to enable debug logs since `--debug` is intuitive enough.

   "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0001] Response #0
< Response Status: "401 Unauthorized"
< Response headers:
   "Content-Length": "73"
   "X-Github-Request-Id": "9FC6:30019C:17C06:1C462:66AD0463"
   "Content-Type": "application/json"
   "Www-Authenticate": "Bearer realm=\"https://ghcr.io/token\",service=\"ghcr.io\",scope=\"repository:oras-project/oras:pull\""
   "Date": "Fri, 02 Aug 2024 16:08:04 GMT" 
DEBU[0001] Request #1
> Request URL: "https://ghcr.io/token?scope=repository%3Aoras-project%2Foras%3Apull&service=ghcr.io"
> Request method: "GET"
> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew" 
DEBU[0002] Response #1
< Response Status: "200 OK"
< Response headers:
   "Content-Type": "application/json"
   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 
```

**Suggested changes:**

```

> $ oras copy ghcr.io/oras-project/oras:v1.2.0 --to-oci-layout oras-dev:v1.2.0 --verbose debug

Reviewer 5: this would be a NACK from me and I'm honestly in favor or removing debug as flag.

```

```
2024-08-02 23:56:02 > msg=Request #0
> Request URL: "https://ghcr.io/v2/oras-project/oras/manifests/v1.2.0"
> Request method: "GET"
> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew"
   "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 

> 

Reviewer: Do we want to insert separators like `-----`?

Reviewer 5: Would recommend looking at other tools. There are so many nicer options - 
![image](https://github.com/user-attachments/assets/3a8967a2-9210-46c7-8a51-458cdbf3f4b7)

Author: I propose to add separator lines between each request and response for readability. This log separator format seems much more common.

Reviewer 3: Maybe an empty line between incoming and outgoing message

   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 
<  Response body:
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:42c524c48e0672568dbd2842d3a0cb34a415347145ee9fe1c8abaf65e7455b46",
      "size": 1239,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    ···
}

```

### oras push/pull/attach

> The verbose output of commands `pull`, `push`, `attach` is now printed out in the default (status) output, the `--verbose` is no longer needed. Considering the progress bar is showed on terminal by default, the verbose output should be available on non-terminal environment using `--no-tty`. See `oras pull` command as an example:

Reviewer 6: ```suggestion
The verbose output of commands `pull`, `push`, `attach` is now printed out in the default (status) output, the `--verbose` is no longer needed. Considering the progress bar is shown on terminal by default, the verbose output should be available on non-terminal environment using `--no-tty`. See `oras pull` command as an example:
```

- **Purpose**: Debug logs are specifically aimed to facilitate ORAS developers to diagnose ORAS tool itself. They contain detailed technical information that is useful for troubleshooting problems.
- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of a possible issue with the tool itself.
- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.

### Common Conventions

Here are the common conventions to print clear and analyzable debug logs.

### **Timestamp Each Log Entry**
- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.

>   - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`

Reviewer 2: Usually, timestamp is the first field. Other issue is the precision. This example shows the precision of 1 second. However, it is not accurate for most cases.

Author: It should be the first field. Fixed the timestamp order. 

@shizhMSFT  Would you mind clarifying why use second-based precision to show the timestamp and debug logs are not accurate for most cases?

Reviewer: We can use the [`RFC3339Nano`](https://pkg.go.dev/time#pkg-constants) format, which is more precise. Example: `2024-11-12T20:40:58.6738192Z`.

Author: Updated per @shizhMSFT 's suggestion

- **Privacy and Security:** Abstain from logging sensitive information such as passwords, personal data, or authentication tokens.
  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]`  (exclude authentication token and password information).

## Proposals for ORAS CLI

Based on the concepts and conventions above, here are the proposal for ORAS diagnose experience improvement: 

- Deprecate the `--verbose` flag and keep `--debug` flag to avoid ambiguity. It is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.
- Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 
- Add two empty lines as the separator between each request and response for readability.
- Add timestamp of each request and response to the beginning of each request and response.

> - Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.

Reviewer: Do we print out the response body upon success only?  Since we should print the error code as well, should we print the response body when the media type is JSON or text, no matter success or failure?

> - Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.

Reviewer 2: It is not "the `mediaType` of the processed OCI object" but the `Content-Type` of the HTTP response body.

Reviewer 2: Details can be found in [RFC 2616 Hypertext Transfer Protocol -- HTTP/1.1](https://www.rfc-editor.org/rfc/rfc2616).

Reviewer 2: See also [Media Types registered in IANA](https://www.iana.org/assignments/media-types/media-types.xhtml)

Author: Fixed

> - Upon success, print out the response body in the debug logs if the `mediaType` of the processed OCI object (e.g. image manifest) is JSON or plain text format. The [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.

Reviewer 2: Considering the security, what's the size limit of printing the response body?

Author: @shizhMSFT 

Size limit is a valid concern as it is crucial for preventing ORAS clients from overloading the registry server with massive payloads that could lead to performance issues. 

However, I don't have a certain value for the size limit on the ORAS client side. Is it depends on the OCI registry? According to the [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pushing-manifests), should ORAS determines the maximum size of the response body based on the specific registry's implementation?

> A registry SHOULD enforce some limit on the maximum manifest size that it can accept. A registry that enforces this limit SHOULD respond to a request to push a manifest over this limit with a response code 413 Payload Too Large.

Reviewer 2: We can decide a value by ourselves.

Author: Revised this sentence accordingly.

## Investigation on other CLI tools

To make sure the ORAS diagnose functions are natural and easier to use, it worth knowing how diagnose functions work in other popular client tools. 

#### Curl

Curl only has a `--verbose` option to output verbose logs. No `--debug` option.

#### Docker and Podman

Docker provides two options `--debug` and `--log-level`  to control debug logs output within different log levels, such as INFO, DEBUG, WARN, etc. No `--verbose` option. Docker has its own daemon service running in local so its logs might be much more complex.

#### Helm

Helm CLI tool provides a global flag `--debug` to enable verbose output.

## Examples in ORAS

This section lists the current behaviors of ORAS debug logs, proposes the suggested changes to ORAS CLI commands. 

### oras copy

Pick the first two requests and responses as examples:

```
oras copy ghcr.io/oras-project/oras:v1.2.0 --to-oci-layout oras-dev:v1.2.0 --debug
```

> **Current debug log**

Reviewer: Should we call out that this logs are for TTY and mention the current logs in non-TTY mode as well?

Author: @Wwwsylvia  I mentioned the tiny difference of debug logs between the TTY and non-TTY modes in the Proposals section. Does this description add more clarity? 

> Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, **the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on terminal but keeping plain text in a non-TTY mode respectively.**

Reviewer: I meant we can put examples of the current logs for both TTY and non-TTY.

Reviewer: An example of logs for non-TTY can be get from `oras repo ls localhost:5000 --debug 2>&1 | cat`

Reviewer: Logs for non-TTY looks like:

> time=2024-10-10T18:45:10+08:00 level=debug msg=Request #0
> > Request URL: "http://localhost:5000/v2/240923test1/manifests/v1"
> > Request method: "HEAD"

Author: @Wwwsylvia  The debug logs `--debug` implies non-TTY mode, which is the same as the output with using `--no-tty`. How to print out the debug logs in TTY mode?

Reviewer: `oras repo ls localhost:5000 --debug` prints logs in TTY mode

Author: Added TTY and non-TTY debug logs per suggestions

      "digest": "sha256:42c524c48e0672568dbd2842d3a0cb34a415347145ee9fe1c8abaf65e7455b46",
      "size": 1239,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    ···
}

```

### oras push/pull/attach

The verbose output of commands `pull`, `push`, `attach` is now printed out in the default (status) output, the `--verbose` is no longer needed. Considering the progress bar is shown on terminal by default, the verbose output should be available on non-terminal environment using `--no-tty`. See `oras pull` command as an example:

```bash
$ oras push --oci-layout layout-test:sample README.md --no-tty

Preparing README.md
Uploading 9500d720111f README.md
Uploading 44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  44136fa355b3 application/vnd.oci.empty.v1+json
Uploaded  9500d720111f README.md
Uploading 655f5cc5d5d2 application/vnd.oci.image.manifest.v1+json
Uploaded  655f5cc5d5d2 application/vnd.oci.image.manifest.v1+json
Pushed [oci-layout] layout-test:sample
ArtifactType: application/vnd.unknown.artifact.v1
Digest: sha256:655f5cc5d5d2e7ff2ab90378514103996523b065ce5ef43cd6e6a4de7d80a535
```

> ### Show user's environment details

Reviewer: Looks like this is not included in the Proposal section?

Author: Updated

Specifically, there are existing GitHub issues raised in the ORAS community.

- The user is confused about when to use `--verbose` and `--debug`. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).
- Poor readability of debug logs. No separator lines between request and response information. Users need to add separator lines manually for readability. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).
- Critical information is missing in debug logs. For example, the [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and metadata of the processed resource object (e.g. image manifest) are not displayed.
- The detailed operation information is missing in verbose output. For example, how many and where are objects processed. Less or no verbose output of ORAS commands in some use cases.
- Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

## Concepts

At first, the output of ORAS flag `--verbose` and `--debug` should be clarified before restructuring them. 

### Output

> There are three types of output in ORAS CLI:

Reviewer 2: Logs are logs, and logs may not be displayed or returned. They may be collected in `systemd`, or sent to `fluentd`.

- **Status output**: such as progress information, progress bar in pulling or pushing files.
- **Metadata output**: showing what has been pulled (e.g. filename, digest, etc.) in specified format, such as JSON, text.
- **Content output**: it is to output the raw data obtained from the remote registry server or file system, such as the pulled artifact content save as a file.
- **Error output**: error message are expected to be helpful to troubleshoot where the user has done something wrong and the program is guiding them in the right direction.

The target users of these types of output are standard users. Currently, the output of ORAS `--verbose` flag only exists in oras `pull/push/attach/discover` commands, which prints out detailed status output and metadata output. 

It is intended for end-users who want to observe the detailed file operation when using ORAS. It gives users a comprehensive view of what the tool is doing at every step and how long does it take when push or pull a file.

Since ORAS v1.2.0, progress bar is enabled in `pull/push/attach` by default, thus the ORAS output is already verbose in a terminal.

### Logs

> Logs focus on providing technical details for in-depth diagnosing and troubleshooting issues. It is intended for developers or technical users who need to understand the inner workings of the tool. Debug logs are detailed and technical, often including HTTP request and response from interactions between client and server, as well as code-specific information. In general, there are different levels of log including `DEBUG`, `INFO`, `WARNING`, `ERROR`, but only `DEBUG` level log is used in ORAS, which is controlled by the flag `--debug`. 

Reviewer 2: Some log implementation has the `TRACE` level, and the `FATAL` level.

Author: I changed the description as follows:

> In general, there are different levels of logs. [Logrus](https://github.com/sirupsen/logrus) has been used by ORAS, which has seven logging levels: `Trace`, `Debug`, `Info`, `Warning`, `Error`, `Fatal` and `Panic`, but only `DEBUG` level log is used in ORAS, which is controlled by the flag `--debug`.

- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.

## Proposals for ORAS CLI

- Deprecate the global flag `--verbose` and only remain `--debug` to avoid ambiguity. Based on the concept above, it is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs in ORAS as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
- Introduce a new flag `--detail` to replace the existing global flag `--verbose` of commands like `pull`, `push`, `attach`, and `discover` for detailed output.
- Add separator lines between each request and response for readability.

> - Add timestamp of each request and response to the beginning of each request and response.

Reviewer 2: We probably just need to adjust the logrus implementation. /cc @Wwwsylvia for feasibility assessment.

> - Add the response body including [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and the metadata of processed OCI object (e.g. image manifest) to the debug logs

Reviewer 2: What does it mean? Is it applicable to OCI image layout?

Reviewer 2: Error code is in the response body as we do extract error code from the JSON response body.

> - Define guiding principles and conventions for writing clear and conductive debug logs

Reviewer 2: What's the purpose / motivation of this?

Author: Removed the "common guiding principles" and only reserved two conventions in this spec

### Guiding Principles

> Here are the guiding principles to write debug logs.

Reviewer 3: I would suggest we can just select a popular logging library or tool and follow the formats that are defined by the library or tool, because I don't think it is worth spending time to develop/customize our own logging formats, unless it is easy to customize using the library or tool.

Reviewer 2: I agree. Currently, we use [logrus](https://github.com/sirupsen/logrus), which is used by docker daemon / CLI. There are also other log implementations like [zap](https://github.com/uber-go/zap), [apex](https://github.com/apex/log), [slog](https://pkg.go.dev/log/slog), etc.

Author: Agree. Removed the "common guiding principles" and only reserved two conventions in this spec

> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew" 

2024-08-02 23:56:03 < msg=Response #1
< Response Status: "200 OK"
< Response headers:
   "Content-Type": "application/json"
   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 
<  Response body:
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:42c524c48e0672568dbd2842d3a0cb34a415347145ee9fe1c8abaf65e7455b46",
      "size": 1239,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    ···
}

```

### Show user's environment details

Output the user's running environment details of ORAS such as operating system and architecture information would be helpful to help the ORAS developers locate the issue and reproduce easier. 

For example, the operating system and architecture are supposed to be outputted in `oras version`: 

```bash
$ oras version

ORAS Version:    1.2.0+Homebrew
Go version:      go1.22.3

> OS/Arch:         linux/amd64

Reviewer 2: Missing git information

Author: Updated

      "digest": "sha256:42c524c48e0672568dbd2842d3a0cb34a415347145ee9fe1c8abaf65e7455b46",
      "size": 1239,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    ···
}

```

## Open Questions

1. Should ORAS applies appropriate log levels to differentiate the log inforamtion? For example:

- **Debug Level:** Reserve the `DEBUG` log level for detailed, technical information meant for developers.
  - Example: `DEBUG: Parsed manifest with 3 layers. Digest: sha256:abcd1234`
  
- **Other Levels:** Use other log levels (Info, Warn, Error) to avoid cluttering debug logs with less vital information.
  - Example (Info): `INFO: Successfully pushed artifact oras-demo:v1`
  - Example (Error): `ERROR: Failed to push artifact oras-demo:v1 due to network timeout`

2. Is it much more reasonable to output the user's environment details in `oras version` rather than in debug logs? The details about the environment where the tool is running, such as OS version and architecture, tool version, and environment variables (excluding sensitive data) would be super helpful to help ORAS developers and users to diagnose an issue.

> For example, the command and output will be as follows if output these information in `oras version`: 

Reviewer 5: I would avoid this for now and have the users provide environment information themselves when reporting the issue.

- Introduce a new flag `--detail` to replace the existing global flag `--verbose` of commands like `pull`, `push`, `attach`, and `discover` for detailed output.
- Add separator lines between each request and response for readability.
- Add timestamp of each request and response to the beginning of each request and response.
- Add the response body including [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and the metadata of processed OCI object (e.g. image manifest) to the debug logs
- Define guiding principles and conventions for writing clear and conductive debug logs

### Guiding Principles

Here are the guiding principles to write debug logs.

### 1. **Timestamp Each Log Entry**
- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.

>   - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`

Reviewer 6: Was it discussed at all to put the timestamp before the log level 
```suggestion
  - Example: `[2023-10-01T12:00:00Z] DEBUG:  Starting metadata retrieval for repository oras-demo`
```

Reviewer: We discussed this in [this comment](https://github.com/oras-project/oras/pull/1483#discussion_r1771208683).
Basically `logurs` supports something like

> DEBU[2024-09-23T18:56:49+08:00] Request #0

Reviewer 6: Personal preference, but I like it when the timestamp is first. Easier to manually/visually grep.

Author: I personally prefer put the timestamp before the log level since it has better readability. Is this feasible in `logurs`? @Wwwsylvia

Reviewer: It is doable, just that we need to implement a custom formatter. But we will need a custom formatter anyway to meet other requirements.

- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.
- Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 
- Add two empty lines as the separator between each request and response for readability.
- Add timestamp of each request and response to the beginning of each request and response.
- Print out the response body in the debug logs if the [Content-Type](https://www.rfc-editor.org/rfc/rfc2616#section-14.17) of the HTTP response body is JSON or plain text format. ORAS SHOULD limit the size of a response body for preventing clients from overloading the registry server with massive payloads that could lead to performance issues.
- Upon a response with failures, the [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) should be included in the response body by default for diagnose purposes.

> - Show running environment details of ORAS such as `OS/Arch` in the output of `oras version`. It would be helpful to help the ORAS developers locate and reproduce the issue easier. 
> - Summarize common conventions for writing clear and analyzable debug logs.
> - Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on terminal but keeping plain text in a non-TTY mode respectively.

Reviewer: For better readability of the spec, consider changing the order of line 79 and line 81, so that line 75-79 are all about logging format and line 81 is about ORAS version.

```suggestion
- Considering ORAS is not a daemon service so parsing debug logs to a logging system is not a common scenario. The target users of the debug logs are normal users and ORAS developers. Thereby, the debug logs in TTY mode and non-TTY (`--no-tty`) should be consistent, except for the color. Specifically, debug logs SHOULD be colored-code in a TTY mode for better readability on terminal but keeping plain text in a non-TTY mode respectively.
- Summarize common conventions for writing clear and analyzable debug logs.
- Show running environment details of ORAS such as `OS/Arch` in the output of `oras version`. It would be helpful to help the ORAS developers locate and reproduce the issue easier. 
```

Author: Updated

3. List the guiding principles to write comprehensive, clear, and conducive debug output and debug logs for effective diagnosis.
4. Propose solutions to improve the diagnostic experience for ORAS CLI users and developers.

## Problem Statement

Specifically, there are existing GitHub issues raised in the ORAS community.

- The user is confused about when to use `--verbose` and `--debug`. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).
- Poor readability of debug logs. No separator lines between request and response information. Users need to add separator lines manually for readability. See the relevant issue [#1382](https://github.com/oras-project/oras/issues/1382).
- Critical information is missing in debug logs. For example, the [error code](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#error-codes) and metadata of the processed resource object (e.g. image manifest) are not displayed.
- The detailed operation information is missing in verbose output. For example, how many and where are objects processed. Less or no verbose output of ORAS commands in some use cases.

> - Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

Author: Based on this example, it looks like timestamp is not visible to users in debug logs by default. Users have to output the timestamp information with an additional command `cat`. 

How about making timestamp information to each request and response by default?

> - Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

Reviewer 2: It depends on logrus. @Wwwsylvia Could you provide options in logrus?

> - Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

Reviewer: Just tested a bit. We can enable [`FullTimestamp`](https://pkg.go.dev/github.com/sirupsen/logrus#TextFormatter) to achieve the goal:
```diff
logger.SetFormatter(&logrus.TextFormatter{
		DisableQuote:  true,
+		FullTimestamp: true,
})
```

As a result, logs in TTY will show as:

> DEBU[2024-09-23T18:56:49+08:00] Request #0

And logs in non-TTY will show as:

> time=2024-09-23T18:59:56+08:00 level=debug msg=Request #0

> - Timestamp of each request and response is missing in debug logs, which is hard to trace historical operation and trace the sequence of events accurately.

Author: Thus, it is technically feasible to make full timestamp information to each request and response by default.

## Concepts

At first, the output of ORAS flag `--verbose` and `--debug` should be clarified before restructuring them. 

### Output

There are three types of output in ORAS CLI:

- **Status output**: such as progress information, progress bar in pulling or pushing files 
- **Metadata output**: showing what has been pulled (e.g. filename, digest, etc.) in specified format, such as JSON, text

> - **Content output**: it is to output the raw data obtained from the remote registry server or file system, such as the pulled artifact content save as a file

Author: Added error output

> The target users of these types of output are normal users. Currently, the output of ORAS `--verbose` flag belongs to status output since it contains detailed progress status. The output of `--verbose` only exists in oras `pull/push/attach` commands. 

Author: Good catch. I revised this paragraph to include `oras discover`.

- **Purpose**: Debug logs are specifically aim to facilitate ORAS developers to diagnose ORAS tool itself. They contain detailed technical information that is useful for troubleshooting problems.
- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of a possible issue with the tool itself.
- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.

## Proposals for ORAS CLI

- Deprecate the global flag `--verbose` and only remain `--debug` to avoid ambiguity. Based on the concept above, it is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs in ORAS as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.

> - The existing output of `--verbose` in several ORAS commands `pull/push/attach/discover` can still be reserved. Introducing an additional flag to these commands.

Reviewer: To make it clearer, how about something like:

> Introduce a new flag to replace the existing `--verbose` flag for detailed output of commands like `pull`, `push`, `attach`, and `discover`

Author: Agree. I used your suggested words.

- Define guiding principles and conventions for writing clear and conductive debug logs

### Guiding Principles

Here are the guiding principles to write debug logs.

### 1. **Timestamp Each Log Entry**
- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.
  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`

### 2. **Capture API-Specific Details**
- **API Requests:** Log detailed information about API requests made to the registry server, including the HTTP method, endpoint, headers, and body (excluding sensitive information).
  - Example: `DEBUG: [HTTPRequest] POST /v2/oras-demo/blobs/uploads/ Headers: {Content-Length: 524288}`
  
- **API Responses:** Log details about the API responses received, including status codes, headers, and response body (excluding sensitive information).
  - Example: `DEBUG: [HTTPResponse] Status: 201 Created, Headers: {Location: /v2/oras-demo/blobs/uploads/uuid}, Body: {}`

> ### 3. **Log Before and After Critical Operations**

Reviewer 3: I would suggest just removing "critical". Once `--debug` is used, there will be logs output on the console after a command is executed.

> ### 3. **Log Before and After Critical Operations**

Author: I changed `Critical` to `Each` to avoid ambiguity. Theoretically, debug log should capture each API operation and request for diagnose. Is it much more accurate?

- **Operation Logs:** Log before performing critical operations and after completing them, including success or failure status.
  - Example: `DEBUG: Starting upload of layer 2 of 3 for repository oras-demo`
  - Example: `DEBUG: Successfully uploaded layer 2 of 3 for repository oras-demo`

- **State Logs:** Log important state information, function names, important variable state, file name and line number before and after key operations or decisions.

>   - Example: `DEBUG: Current retry attempt: 1, Max retries: 3`

Author: This guiding principle has been removed.

### 4. **Error and Exception Handling**
- **Catch and Log Exceptions:** Always catch exceptions and log them with relevant context and stack traces.
  - Example: `ERROR: Exception occurred in fetchManifest: Network timeout while accessing /v2/oras-demo/manifests/latest`
  
- **Error Codes:** Include specific error codes to facilitate quick identification and resolution.

>   - Example: `ERROR: [ErrorCode: 504] Network timeout while accessing /v2/oras-demo/manifests/latest`

Author: It's debug log. Should error information of ORAS CLI also be included in the debug logs?

>   - Example: `ERROR: [ErrorCode: 504] Network timeout while accessing /v2/oras-demo/manifests/latest`

Reviewer 2: Currently, `oras` only use debug logs.

>   - Example: `ERROR: [ErrorCode: 504] Network timeout while accessing /v2/oras-demo/manifests/latest`

Author: Removed this guiding principle

### 5. **Include Actionable Information**
- **Guidance:** Where possible, provide suggestions within logs about potential fixes or next steps for resolving issues.
  - Example: `DEBUG: [Action] Check network connectivity and retry the operation. Error occurred while accessing /v2/oras-demo/manifests/latest`

- **Diagnostic Tips:** Include information that can assist in diagnosing issues like configuration settings, environment variables, or system states.

>   - Example: `DEBUG: Current registry URL: https://myregistry.io, Timeout setting:

Author: Including actionable information in debug logs can be beneficial if it helps diagnose issues or improve the user's understanding of the tool’s behavior. 

> The only action for users is to present the debug log to our oras developers.

If the user is also a developer, will this design limits the diagnose efficiency in this situation?

>   - Example: `DEBUG: Current registry URL: https://myregistry.io, Timeout setting:

Author: Removed this guiding principle as this should be a part of the error message

### 6. **Avoid Logging Sensitive Information**
- **Privacy:** Abstain from logging sensitive information such as passwords, personal data, or security tokens.
  - Example: `DEBUG: Attempting to authenticate user [UserID: usr123]` (without password details)

- **Compliance:** Ensure that logs adhere to relevant data protection and privacy regulations.
  - Example: Anonymize or obfuscate sensitive customer data in logs.

## Investigation

To make sure the ORAS diagnose functions are natural and easier to use, it worth knowing how diagnose functions work in other popular client tools. 

#### Curl

Curl only has a `--verbose` option to output verbose logs. No `--debug` option.

#### Docker and Podman

Docker provides two options `--debug` and `--log-level`  to control debug logs output within different log levels, such as INFO, DEBUG, WARN, etc. No `--verbose` option. Docker has its own daemon service running in local so its logs might be much more complex.

#### Helm

Helm CLI tool provides a global flag `--debug` to enable verbose output.

#### Kubectl

> Kubectl has a command `kubectl logs` to show logs of objects such as Pod and container. No separate flags for verbose output and debug logs.

Reviewer 2: kubectl is not a good example since all business logic happens at the server side.

Author: Removed kubectl

## Examples in ORAS

This section lists the current behaviors of ORAS debug logs, proposes the suggested changes to ORAS CLI commands. More examples will be appended below.

### oras copy

Pick the first two requests and responses as examples:

```
oras copy ghcr.io/oras-project/oras:v1.2.0 --to-oci-layout oras-dev:v1.2.0 --debug
```

> **Current debug log0-=s**

Reviewer: Is it a typo?

Author: Yes. Resolved.

   "Accept": "application/vnd.docker.distribution.manifest.v2+json, application/vnd.docker.distribution.manifest.list.v2+json, application/vnd.oci.image.manifest.v1+json, application/vnd.oci.image.index.v1+json, application/vnd.oci.artifact.manifest.v1+json" 
DEBU[0001] Response #0
< Response Status: "401 Unauthorized"
< Response headers:
   "Content-Length": "73"
   "X-Github-Request-Id": "9FC6:30019C:17C06:1C462:66AD0463"
   "Content-Type": "application/json"
   "Www-Authenticate": "Bearer realm=\"https://ghcr.io/token\",service=\"ghcr.io\",scope=\"repository:oras-project/oras:pull\""
   "Date": "Fri, 02 Aug 2024 16:08:04 GMT" 
DEBU[0001] Request #1
> Request URL: "https://ghcr.io/token?scope=repository%3Aoras-project%2Foras%3Apull&service=ghcr.io"
> Request method: "GET"
> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew" 
DEBU[0002] Response #1
< Response Status: "200 OK"
< Response headers:
   "Content-Type": "application/json"
   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 
```

**Suggested debug logs:**

```

> 2024-08-02 23:56:02 > msg=Request #0

Reviewer 2: See https://github.com/oras-project/oras/pull/1483#discussion_r1759686580

> Request headers:
   "User-Agent": "oras/1.2.0+Homebrew" 

2024-08-02 23:56:03 < msg=Response #1
< Response Status: "200 OK"
< Response headers:
   "Content-Type": "application/json"
   "Docker-Distribution-Api-Version": "registry/2.0"
   "Date": "Fri, 02 Aug 2024 16:08:05 GMT"
   "Content-Length": "69"
   "X-Github-Request-Id": "9FC6:30019C:17C0D:1C46C:66AD0464" 
<  Response body:
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.index.v1+json",
  "manifests": [
    {
      "mediaType": "application/vnd.oci.image.manifest.v1+json",
      "digest": "sha256:42c524c48e0672568dbd2842d3a0cb34a415347145ee9fe1c8abaf65e7455b46",
      "size": 1239,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    ···
}

```

### Show user's environment details

Output the user's running environment details of ORAS such as operating system and architecture information would be helpful to help the ORAS developers locate the issue and reproduce easier. 

For example, the operating system and architecture are supposed to be outputted in `oras version`: 

```bash
$ oras version

ORAS Version:    1.2.0+Homebrew
Go version: go1.22.3

> OS/Arch: Linux AMD64

Reviewer 2: It is better to use the standard os/arch format: `linux/amd64` (i.e. `<OS>/<ARCH>`).

Reviewer 2: Example:
```console
$ docker version
Client:
 Version:           27.2.0
 API version:       1.47
 Go version:        go1.21.13
 Git commit:        3ab4256
 Built:             Tue Aug 27 14:14:20 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Desktop  ()
 Engine:
  Version:          27.2.0
  API version:      1.47 (minimum version 1.24)
  Go version:       go1.21.13
  Git commit:       3ab5c7d
  Built:            Tue Aug 27 14:15:15 2024
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          1.7.20
  GitCommit:        8fc6bcff51318944179630522a095cc9dbf9f353
 runc:
  Version:          1.1.13
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Author: Updated to `<OS>/<ARCH>`

> OS/Arch: Linux AMD64

Reviewer 2: Make it a table

  - Example: `[2024-08-02 23:56:02.6738192Z] Starting metadata retrieval for repository oras-demo`

### **Avoid Logging Sensitive Information**
- **Privacy and Security:** Abstain from logging sensitive information such as passwords, personal data, or authentication tokens.
  - Example: `[2024-08-02 23:56:02.7338192Z] Attempting to authenticate user [UserID: usr123]`  (exclude authentication token and password information).

## Proposals for ORAS CLI

Based on the concepts and conventions above, here are the proposal for ORAS diagnose experience improvement: 

- Deprecate the `--verbose` flag and keep `--debug` flag to avoid ambiguity. It is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs as it is in ORAS. Meanwhile, this change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.
- Make the verbose output of commands `pull`, `push`, `attach` as the default (status) output. See examples at the bottom.

> - Make the verbose output of command  `discover` as a formatted output, controlled by `--format tree-full`. 

Reviewer: One question: Although `--verbose` is deprecated, but it is not removed. Should it still work as before for `oras discover`? That is, printing the same detailed output as before?

BTW, I just realized that there is no example for this.

Author: @Wwwsylvia  The second proposal is to make the verbose output as the default (status) output in commands `pull`, `push`, `attach` , without the `--verbose` flag anymore. Let me append an example in the bottom for clarification. Thanks for catching it.

There are three types of output in ORAS CLI:

- **Status output**: such as progress information, progress bar in pulling or pushing files.
- **Metadata output**: showing what has been pulled (e.g. filename, digest, etc.) in specified format, such as JSON, text.
- **Content output**: it is to output the raw data obtained from the remote registry server or file system, such as the pulled artifact content save as a file.
- **Error output**: error message are expected to be helpful to troubleshoot where the user has done something wrong and the program is guiding them in the right direction.

> The target users of these types of output are standard users. Currently, the output of ORAS `--verbose` flag only exists in oras `pull/push/attach/discover` commands, which prints out detailed status output and metadata output. 

Reviewer: nit: Instead of "standard users", how about "regular users"?

Author: @Wwwsylvia  "General users" or "Regular Users" looks good to me since these terms are commonly used in Linux.  Or we can also consider using "user" and "developer" to differentiate two types of personas.

Reviewer 2: `general users` sounds better.

## Proposals for ORAS CLI

- Deprecate the global flag `--verbose` and only remain `--debug` to avoid ambiguity. Based on the concept above, it is reasonable to continue using `--debug` to enable the output of `DEBUG` level logs in ORAS as it is in ORAS. Meanwhile, This change will make the diagnose experience much more straightforward and less breaking since only ORAS `pull/push/attach/discover` commands have verbose output.

> - Introduce a new flag `--detail` to replace the existing global flag `--verbose` of commands like `pull`, `push`, `attach`, and `discover` for detailed output.

Reviewer: How about `--long`?

Reviewer 3: I don't think `--detail` is a good option, since it will cause similar confusion as `--verbose`. Questions like is it about detailed logs or about detailed other outputs? 

One option is output the `--verbose` content by default, then we do not need any flags for verbose output that are not logs. This also solves the problem that some commands have `--verbose` flags, some don't.

Reviewer 3: Will `--debug` be a global flag, or just available for some commands?

Reviewer 2: Instead of `--detail`, I think we can have `--quiet` for `pull`, `push`, and `attach`.

`discover` is different since the `--verbose` takes effect on the metadata output instead of the status output. In that case, we can just remove `--verbose` and introduce new values for `--format` such as `--format tree-full`. It is also a good time to revisit the outputs of `oras discover`.

Author: Agree with @shizhMSFT . Current verbose output in `pull`, `push`, `attach` is progress information but in `discover` it is metadata output. They are different cases.

Author: Updated this proposal accordingly
